notifications_utils/broadcast_areas/__init__.py

```diff
@@ -0,0 +1,168 @@
+import itertools
+from contextlib import suppress
+from pathlib import Path
+from functools import lru_cache
+import geojson
+from notifications_utils.formatters import formatted_list
+
+from notifications_utils.serialised_model import SerialisedModelCollection
+from notifications_utils.safe_string import make_string_safe_for_id
+
+
+@lru_cache(maxsize=128)
+def load_geojson_file(filename):
+
+    path = Path(__file__).resolve().parent / filename
+
+    geojson_data = geojson.loads(path.read_text())
+
+    if not isinstance(geojson_data, geojson.GeoJSON) or not geojson_data.is_valid:
+        raise ValueError(
+            f'Contents of {path} are not valid GeoJSON'
+        )
+
+    return path.stem, geojson_data
+
+
+class IdFromNameMixin:
+
+    @property
+    def id(self):
+        return make_string_safe_for_id(self.name)
+
+    def __repr__(self):
+        return f'{self.__class__.__name__}(<{self.id}>)'
+
+    def __lt__(self, other):
+        # Implementing __lt__ means any classes inheriting from this
+        # method are sortable
+        return self.id < other.id
+
+
+class GetItemByIdMixin:
+    def get(self, id):
+        for item in self:
+            if item.id == id:
+                return item
+        raise KeyError(id)
+
+
+class BroadcastArea(IdFromNameMixin):
+
+    def __init__(self, feature):
+        self.feature = feature
+
+        for coordinates in self.polygons:
+            if coordinates[0] != coordinates[-1]:
+                # The CAP XML format requires shapes to be closed
+                raise ValueError(
+                    f'Area {self.name} is not a closed shape '
+                    f'({coordinates[0]}, {coordinates[-1]})'
+                )
+
+    def __eq__(self, other):
+        return self.id == other.id
+
+    @property
+    def name(self):
+        for possible_name_key in {
+            'rgn18nm', 'ctyua16nm', 'ctry19nm',
+        }:
+            with suppress(KeyError):
+                return self.feature['properties'][possible_name_key]
+
+        raise KeyError(f'No name found in {self.feature["properties"]}')
+
+    @property
+    def polygons(self):
+        if self.feature['geometry']['type'] == 'MultiPolygon':
+            return [
+                polygons[0]
```

Comment:
does each array within `coordinates ever contain more than 1 thing? i had a quick look in the spec but wasn't clear 🤷‍♂️

Reply:
Do you mean an array that mixes coordinate pairs and other things?

Reply:
i mean we're saying `polygons[0]` - is there ever a `polygons[1]`? (And the same when we're accessing point [0] in Polygons.

Turns out it could hypothetically: https://tools.ietf.org/html/rfc7946#appendix-A.3

> Coordinates of a Polygon are an array of linear ring (see
>   Section 3.1.6) coordinate arrays.  The first element in the array
>   represents the exterior ring.  Any subsequent elements represent
>   interior rings (or holes).

don't think we need to worry about these until we find a data set with one in